cornell-dti/course-plan
# Populate conflict modal with requirements

### Summary 

This pull request hooks up the conflict modal created in #659 to the requirements side of CoursePlan such that conflicts can now be resolved!

![image](https://user-images.githubusercontent.com/25535093/165544309-e1448fb4-1085-4d32-b845-b5dacb3e9e96.png)

- [x] Only open conflicts modal when there are conflicts for that course
- [x] Populate modal with the actual requirements choices (and refactor to use a map)
- [x] Show self checks if a course can apply to them 
- [x] **NEW** Exclude self-checks that are not in conflict with the other non-self-check reqs
- [x] **NEW** Hide self-check warning if there are no self-checks in modal

Remaining TODOs (for the next PR)

- [x] Ensure the choice affects the added course and requirements fulfilled
- [x] Display the confirmation after the conflict modal is closed

### Questions Remaining 

- What should the selectable requirement logic be? Right now self-checks show up if a course can be applied to that self-check at all. This does not handle cases though where it does not cause a conflict - i.e. if you add HD 1150, it can apply to CS self-checks AND one of the minor reqs at the same time). It also means self-check req could show up in multiple conflicts in the same modal, as shown above.
- Should I delete the "major-cs-" or "minor-hd-" before the requirement name? I assume so, but then there will be no indication of major/minor name which could be confusing for broad req names like "Electives."
- Should the warning text "Requirements with icon are..." only show if there are selectable reqs in the conflict? Right now it always shows up.

### Test Plan 
Set major to Human Development (in human ecology), minor to Human Development 
1. Try to add HD 1150 - confirm you get two sets of conflicts, one for the major and one for the minor, as this course can fulfill multiple reqs but only one at a time in each group.
2. **NEW** Confirm the first conflict has all self-check reqs, including college (as only one of the college & major reqs can be selected). Similarly, confirm the minor group has no self-check reqs.

Set major to CS (in engineering), keep Human Development minor.
1. Try to add CS 4152 - confirm there is a conflict between CS 3000+ elective req, practicum req, and 3 selectable reqs.
![image](https://user-images.githubusercontent.com/25535093/163850581-19871fbb-d47e-472c-96d3-b4f856a6d487.png)
3. Try to add HD 1150 - confirm that there is a conflict between the two minor requirements. **NEW** Confirm there are no selectable reqs and the selectable req warning does not show.
![image](https://user-images.githubusercontent.com/25535093/165545411-c088a6cf-67d9-4106-9b37-810cc8682643.png)
4. Try to add CS 1110 - confirm that if a course has no conflicts (but can apply to self-check reqs), the conflict modal is completely skipped and it is just added to your plan.
6. Confirm that if the flag is turned off, the old add modal v2 shows up after a course is selected.



### Notes 

To enable the GK and use this feature, type GK.enableRequirementConflicts() into the console on the branch or hosted link
